src/Microsoft.Azure.WebJobs.Host/Executors/JobHostContextFactory.cs

```diff
@@ -27,121 +27,162 @@
-            TraceWriter trace = new ConsoleTraceWriter(config.Tracing, consoleProvider.Out);
-
-            // Register system services with the service container
-            config.AddService<INameResolver>(nameResolver);
-
-            if (exceptionHandler != null)
-            {
-                exceptionHandler.Initialize(host);
-            }
+            TraceWriter trace = new ConsoleTraceWriter(rawConfig.Tracing, consoleProvider.Out);
+            config.AddService<TraceWriter>(trace);
 
             ExtensionConfigContext context = new ExtensionConfigContext
             {
-                Config = config,
-                Trace = trace                
+                Config = rawConfig,
+                Trace = trace
             };
             InvokeExtensionConfigProviders(context);
 
             if (singletonManager == null)
             {
-                singletonManager = new SingletonManager(storageAccountProvider, exceptionHandler, config.Singleton, trace, hostIdProvider, config.NameResolver);
+                singletonManager = new SingletonManager(storageAccountProvider, exceptionHandler, rawConfig.Singleton, trace, hostIdProvider, config.GetService<INameResolver>());
+                config.AddService<SingletonManager>(singletonManager);
             }
 
             IExtensionRegistry extensions = config.GetExtensions();
             ITriggerBindingProvider triggerBindingProvider = DefaultTriggerBindingProvider.Create(nameResolver,
                 storageAccountProvider, extensionTypeLocator, hostIdProvider, queueConfiguration, blobsConfiguration, exceptionHandler,
                 messageEnqueuedWatcherAccessor, blobWrittenWatcherAccessor, sharedContextProvider, extensions, singletonManager, trace);
+            config.AddService<ITriggerBindingProvider>(triggerBindingProvider);
 
             if (bindingProvider == null)
             {
                 bindingProvider = DefaultBindingProvider.Create(nameResolver, converterManager, storageAccountProvider, extensionTypeLocator, messageEnqueuedWatcherAccessor, blobWrittenWatcherAccessor, extensions);
+                config.AddService<IBindingProvider>(bindingProvider);
             }
+        }
 
-            bool hasFastTableHook = config.GetService<IAsyncCollector<FunctionInstanceLogEntry>>() != null;
-            bool noDashboardStorage = config.DashboardConnectionString == null;
-            if (hasFastTableHook && noDashboardStorage)
+        // Do the runtime intitialization. This happens after the static initialization. 
+        // This mainly means:
+        // - indexing the functions 
+        // - spinning up the listeners (so connecting to the services)
+        private static async Task<JobHostContext> DoRuntimeInitialization(this ServiceProviderWrapper config, JobHostConfiguration rawConfig, JobHost host, CancellationToken shutdownToken, CancellationToken cancellationToken)
```

Comment:
Rename to CreateJobHostContextAsync, which better describes what the method actually does